File: graphlib-3/go/lib/traversal.go
package graphlib

import (
	"sort"
)

// BFS performs breadth-first search from a start node
func BFS(graphID, startNode string) map[string]interface{} {
	g, ok := Store.GetGraph(graphID)
	if !ok {
		return map[string]interface{}{
			"order":  []string{},
			"levels": map[string]int{},
			"parent": map[string]string{},
		}
	}

	if !g.Nodes[startNode] {
		return map[string]interface{}{
			"order":  []string{},
			"levels": map[string]int{},
			"parent": map[string]string{},
		}
	}

	order := []string{}
	levels := make(map[string]int)
	parent := make(map[string]string)
	visited := make(map[string]bool)

	queue := []string{startNode}
	visited[startNode] = true
	levels[startNode] = 0

	for len(queue) > 0 {
		node := queue[0]
		queue = queue[1:]
		order = append(order, node)

		// Get sorted neighbors for deterministic order
		neighbors := make([]string, 0, len(g.Adj[node]))
		for neighbor := range g.Adj[node] {
			neighbors = append(neighbors, neighbor)
		}
		sort.Strings(neighbors)

		for _, neighbor := range neighbors {
			if !visited[neighbor] {
				visited[neighbor] = true
				levels[neighbor] = levels[node] + 1
				parent[neighbor] = node
				queue = append(queue, neighbor)
			}
		}
	}

	return map[string]interface{}{
		"order":  order,
		"levels": levels,
		"parent": parent,
	}
}

// DFS performs depth-first search from a start node
func DFS(graphID, startNode string) map[string]interface{} {
	g, ok := Store.GetGraph(graphID)
	if !ok {
		return map[string]interface{}{
			"order":     []string{},
			"discovery": map[string]int{},
			"finish":    map[string]int{},
			"parent":    map[string]string{},
		}
	}

	if !g.Nodes[startNode] {
		return map[string]interface{}{
			"order":     []string{},
			"discovery": map[string]int{},
			"finish":    map[string]int{},
			"parent":    map[string]string{},
		}
	}

	order := []string{}
	discovery := make(map[string]int)
	finish := make(map[string]int)
	parent := make(map[string]string)
	visited := make(map[string]bool)
	time := 0

	var dfsVisit func(node string)
	dfsVisit = func(node string) {
		visited[node] = true
		discovery[node] = time
		time++
		order = append(order, node)

		// Get sorted neighbors for deterministic order
		neighbors := make([]string, 0, len(g.Adj[node]))
		for neighbor := range g.Adj[node] {
			neighbors = append(neighbors, neighbor)
		}
		sort.Strings(neighbors)

		for _, neighbor := range neighbors {
			if !visited[neighbor] {
				parent[neighbor] = node
				dfsVisit(neighbor)
			}
		}

		finish[node] = time
		time++
	}

	dfsVisit(startNode)

	return map[string]interface{}{
		"order":     order,
		"discovery": discovery,
		"finish":    finish,
		"parent":    parent,
	}
}

File: graphlib-1/py/graphlib/traversal.py
"""BFS and DFS traversal implementations."""
from typing import Dict, List, Any
from collections import deque
from .types import get_graph


def bfs(graph_id: str, start_node: str) -> Dict[str, Any]:
    """Breadth-first search traversal."""
    graph = get_graph(graph_id)
    if graph is None:
        return {"order": [], "levels": {}, "parent": {}}
    if start_node not in graph.nodes:
        return {"order": [], "levels": {}, "parent": {}}

    order = []
    levels = {}
    parent = {}
    visited = set()

    queue = deque([start_node])
    visited.add(start_node)
    levels[start_node] = 0

    while queue:
        node = queue.popleft()
        order.append(node)

        # Get neighbors in sorted order for deterministic traversal
        neighbors = sorted(graph.adj.get(node, {}).keys())
        for neighbor in neighbors:
            if neighbor not in visited:
                visited.add(neighbor)
                levels[neighbor] = levels[node] + 1
                parent[neighbor] = node
                queue.append(neighbor)

    return {"order": order, "levels": levels, "parent": parent}


def dfs(graph_id: str, start_node: str) -> Dict[str, Any]:
    """Depth-first search traversal."""
    graph = get_graph(graph_id)
    if graph is None:
        return {"order": [], "discovery": {}, "finish": {}, "parent": {}}
    if start_node not in graph.nodes:
        return {"order": [], "discovery": {}, "finish": {}, "parent": {}}

    order = []
    discovery = {}
    finish = {}
    parent = {}
    visited = set()
    time = [0]  # Use list to allow modification in nested function

    def dfs_visit(node: str) -> None:
        visited.add(node)
        discovery[node] = time[0]
        time[0] += 1
        order.append(node)

        # Get neighbors in sorted order for deterministic traversal
        neighbors = sorted(graph.adj.get(node, {}).keys())
        for neighbor in neighbors:
            if neighbor not in visited:
                parent[neighbor] = node
                dfs_visit(neighbor)

        finish[node] = time[0]
        time[0] += 1

    dfs_visit(start_node)

    return {"order": order, "discovery": discovery, "finish": finish, "parent": parent}
